### PR TITLE
fix: [CDS-75775]: Export variables with single quotes

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
Currently in command step when we are exporting variables we are wrapping them using double quotes which causes to escape values like $1 in the string, with this change we will quote them using single quotes which will not modify the value for that variable.

Screenshot 2023-08-15 at 1 21 47 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)

### CODE REVIEW

**Review by shivagowda — COMMENTED** [review]
pls unit tests for the newly added code.

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 866 at r1](https://reviewable.io/reviews/harness/harness-core/51573#gh-1295135626:-Nbw_qFK40A5XF9-hVPr:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ea8ada23fbb5423f820563fdf820288766da2a10/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L866)):*
<details><summary><i>Previously, shivagowda (Shivakumar Ningappa) wrote…</i></summary><blockquote>

pls add description and team name for this new FF
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 6 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by ivan-83 — COMMENTED** [review]
LGTM just minor comments. 

> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573#-:-NbyeqQoERE_VMaeSixG:b2oub26)* status: 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/resources/commandtemplates/execlauncherv3.sh.ftl` line 368 at r2](https://reviewable.io/reviews/harness/harness-core/51573#-NbyeQxO9iUm1zQRejj3:-NbyeQxO9iUm1zQRejj4:bufwywg) ([raw file](https://github.com/harness/harness-core/blob/b3510e957ab06a34ce004c69fd1f25cc37b4cb0f/930-delegate-tasks/src/main/resources/commandtemplates/execlauncherv3.sh.ftl#L368)):*
> ```FreeMarker
> # Set Environment Variables.
> <#list envVariables?keys as envVariable>
> export ${envVariable}='${envVariables[envVariable]}'
> ```

@sarthakkasat just to check, you haven't changed the rest of the file just the logic related to env variables quotes, right? 
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/SshCommandTaskParameters.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/51573#-NbyflkDB7-xDrn4aSK6:-NbyflkDB7-xDrn4aSK7:b-bg2n2y) ([raw file](https://github.com/harness/harness-core/blob/b3510e957ab06a34ce004c69fd1f25cc37b4cb0f/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/SshCommandTaskParameters.java#L30)):*
> ```Java
>   SshInfraDelegateConfig sshInfraDelegateConfig;
>   String host;
>   @Builder.Default boolean exportVariableWithSingleQuotes = false;
> ```

Minor - exportVariableWithSingleQuotes is always false when we make an object with the builder or the new constructor.

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/SshCommandTaskParameters.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/51573#-NbyflkDB7-xDrn4aSK6:-NbzJcGSEWIfYJ4DgshV:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/b3510e957ab06a34ce004c69fd1f25cc37b4cb0f/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/shell/SshCommandTaskParameters.java#L30)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Minor - exportVariableWithSingleQuotes is always false when we make an object with the builder or the new constructor.
</blockquote></details>

Done

**Review by sarthakkasat — COMMENTED** [review]
UT's are not required here, as the changes are in internal methods, instead of adding FF we will add it in a setting which will enable to enable and disable evaluating as some users might want the old behaviour.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573#-:-NbzJqqM4W_B53SgFZXT:bjr54lv)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java [hunk 1]
```diff
@@ -192,7 +193,19 @@ private String generateExecLauncherV2Command(SshCommandTaskParameters taskParame
                                                .put("scriptWorkingDirectory", workingDir)
                                                .put("includeTailFunctions", includeTailFunctions)
                                                .build();
-      SshInitCommandTemplates.getTemplate(EXECLAUNCHERV2_SH_FTL).process(templateParams, stringWriter);
+      String templateFileLocation =
+          taskParameters.isEvaluateExportVariable() ? EXECLAUNCHERV2_SH_FTL : EXECLAUNCHERV3_SH_FTL;
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java:197]: constante on left side of equals

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java [hunk 2]
```diff
@@ -192,7 +193,19 @@ private String generateExecLauncherV2Command(SshCommandTaskParameters taskParame
...
+      if (templateFileLocation.equals(EXECLAUNCHERV3_SH_FTL)) {
+        Map<String, String> envVariablesMap = (Map<String, String>) templateParams.get("envVariables");
+        if (envVariablesMap != null) {
+          for (Map.Entry<String, String> entry : envVariablesMap.entrySet()) {
+            String originalValue = entry.getValue();
+            String modifiedValue = originalValue.replace("'", "'\\''");
```
Comment (yogesh-chauhan) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java:203]: extract out to lets say "escape Single Quotes"

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: Some of the checks are too old. 5 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java` line 197 at r5](https://reviewable.io/reviews/harness/harness-core/51573#gh-1301432109:-NcRsUo9DA6kBB9EsUTh:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/35bdc68ce50792b12a860a800b4d1c7519afbd92/930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java#L197)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

constante on left side of equals
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java` line 203 at r5](https://reviewable.io/reviews/harness/harness-core/51573#gh-1301442953:-NcRsUVE6i-wvlY8nH_X:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/35bdc68ce50792b12a860a800b4d1c7519afbd92/930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/ssh/SshInitCommandHandler.java#L203)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

extract out to lets say "escape Single Quotes"
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/resources/commandtemplates/execlauncherv3.sh.ftl` line 368 at r2](https://reviewable.io/reviews/harness/harness-core/51573#-NbyeQxO9iUm1zQRejj3:-NcRs8tYE6G7wF3_t-l9:b2nhswy) ([raw file](https://github.com/harness/harness-core/blob/b3510e957ab06a34ce004c69fd1f25cc37b4cb0f/930-delegate-tasks/src/main/resources/commandtemplates/execlauncherv3.sh.ftl#L368)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

@sarthakkasat just to check, you haven't changed the rest of the file just the logic related to env variables quotes, right?
</blockquote></details>

Yes

**Comment by sarthakkasat** [comment]
trigger commonchecks

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 3]
```diff
@@ -863,6 +863,7 @@ public enum FeatureName {
   CDS_STORE_INSTANCE_STATS_ITERATOR_RUN_TIME(
       "FF for storing when the instance stats iterator ran for a particular service", HarnessTeam.CDP),
   CDS_DISABLE_JIRA_SERVICENOW_RETRY_INTERVAL("For removing usage of retry interval field for ", HarnessTeam.CDC),
+  CDS_DISABLE_EVALUATE_EXPORT_VARIABLES("For exporting variables without evaluating in command step", SPG),
```
Comment (yogesh-chauhan) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:866]: team can be CDS itself

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java [hunk 4]
```diff
@@ -138,6 +139,7 @@ public class SshCommandStepHelper extends CDStepHelper {
   @Inject private SshWinRmArtifactHelper sshWinRmArtifactHelper;
   @Inject private CDExpressionResolver cdExpressionResolver;
   @Inject private NGSecretServiceV2 ngSecretServiceV2;
+  @Inject private NGSettingsClient settingsClient;
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java:142]: this is not needed right ?

**Review by yogesh-chauhan — APPROVED** [review]

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 6 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java` line 142 at r6](https://reviewable.io/reviews/harness/harness-core/51573#gh-1301491769:-NcRzXw_8mIhwHeCgy_s:ba3dfob) ([raw file](https://github.com/harness/harness-core/blob/a0bb3d4583e37a97900aef2d4f851d9139fb2c00/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java#L142)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this is not needed right ?
</blockquote></details>

Removed

**Review by yogesh-chauhan — DISMISSED** [review]

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 6 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 866 at r6](https://reviewable.io/reviews/harness/harness-core/51573#gh-1301491448:-NcS-NpJ6Us8GuZ0mwlN:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a0bb3d4583e37a97900aef2d4f851d9139fb2c00/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L866)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

team can be CDS itself
</blockquote></details>

Done.

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 2 of 6 files at r6, 3 of 3 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573#-:-NcS30MuE0VwLFfsD65Z:bnfp4nl)* status: 6 file(s) to review. 6. unresolved discussions. 1 of 2 approvals obtained.

**Comment by sarthakkasat** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 6 files at r6, 3 of 3 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51573)* status: 2. unresolved discussions.

**Comment by sarthakkasat** [comment]
trigger review

**Comment by sarthakkasat** [comment]
trigger uts

**Comment by sarthakkasat** [comment]
trigger ut4, ut9

**Pull request merged**